### PR TITLE
[rush] Support execution of lifecycle scripts containing Unix slashes on Windows OS

### PR DESCRIPTION
**Pull request by pgonzal**
When running a command from the "scripts" block in package.json, if the command contains Unix-style path slashes, the package managers will convert the slashes to backslashes for the Windows OS.  This is a complicated heuristic.  For example, we want to convert `"node_modules/bin/this && ./scripts/that"` to `"node_modules\bin\this && .\scripts\that"` but we don't want to convert the slashes in  `"cmd.exe /c blah"`.  NPM and PNPM use npm-lifecycle for this, but it unfortunately has a dependency on the entire node-gyp kitchen sink. Yarn has a simplified implementation in fix-cmd-win-slashes.js, but it's not exposed as a library.

Since the "&&" and quoting edge cases are mostly deprecated for Rush projects, we will just do something very basic for now.  We can improve it later if someone actually complains.

### CODE REVIEW

File: apps/rush-lib/src/cli/taskRunner/ProjectTask.ts [hunk 1]
```diff
@@ -126,9 +126,14 @@ export default class ProjectTask implements ITaskDefinition {
         }
 
         // Run the task
-        writer.writeLine(taskCommand);
+
+        const convertedTaskCommand: string = process.platform === 'win32'
```
Comment (iclanton) [inline apps/rush-lib/src/cli/taskRunner/ProjectTask.ts]: >convertedTaskCommand [](start = 14, length = 20)

`normalizedTaskCommand`?

File: apps/rush-lib/src/cli/taskRunner/ProjectTask.ts [hunk 2]
```diff
@@ -243,3 +248,43 @@ function _areShallowEqual(object1: Object, object2: Object, writer: ITaskWriter)
...
+ * but it unfortunately has a dependency on the entire node-gyp kitchen sink.
+ * Yarn has a simplified implementation in fix-cmd-win-slashes.js, but it's not
+ * exposed as a library.
+ *
+ * Since the "&&" and quoting edge cases are mostly deprecated for Rush projects,
+ * we will just do something very basic for now.  We can improve it later if someone
+ * actually complains.
+ */
+export function convertSlashesForWindows(command: string): string {
+  // Match everything up to the first space, "&" or quote
```
Comment (iclanton) [inline apps/rush-lib/src/cli/taskRunner/ProjectTask.ts]: >Match everything up to the first space, "&" or quote [](start = 5, length = 52)

*Match everything up to the first space, "&" or quote in the first capture group.

File: apps/rush-lib/src/cli/taskRunner/ProjectTask.ts [hunk 3]
```diff
@@ -243,3 +248,43 @@ function _areShallowEqual(object1: Object, object2: Object, writer: ITaskWriter)
...
+ * the slashes in  "cmd.exe /c blah".  NPM and PNPM use npm-lifecycle for this,
+ * but it unfortunately has a dependency on the entire node-gyp kitchen sink.
+ * Yarn has a simplified implementation in fix-cmd-win-slashes.js, but it's not
+ * exposed as a library.
+ *
+ * Since the "&&" and quoting edge cases are mostly deprecated for Rush projects,
+ * we will just do something very basic for now.  We can improve it later if someone
+ * actually complains.
+ */
+export function convertSlashesForWindows(command: string): string {
```
[the remark this answers is not in the document]
Comment (pgonzal) [reply apps/rush-lib/src/cli/taskRunner/ProjectTask.ts]: I gave a very long explanation above about why a correct implementation is a lot of work, and why I don't want to do that work.

---
In reply to: [179327831](https://github.com/Microsoft/web-build-tools/pull/601#discussion_r179327831) [](ancestors = 179327831)

File: apps/rush-lib/src/cli/taskRunner/ProjectTask.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (pgonzal) [reply apps/rush-lib/src/cli/taskRunner/ProjectTask.ts]: The reason to do that would be to support spaces in the path:

"node_modules/.bin/my tool"

"Don't do that" is my answer for now heheh.  But seriously, parsing quotes is wandering into the dicey intersection of wildly different shell languages.  My goal here is to solve a small set of use cases that I think we need, and try hard not to break anything else.

---
In reply to: [179327756](https://github.com/Microsoft/web-build-tools/pull/601#discussion_r179327756) [](ancestors = 179327756)

File: apps/rush-lib/src/cli/taskRunner/ProjectTask.ts [hunk 4]
```diff
@@ -243,3 +248,43 @@ function _areShallowEqual(object1: Object, object2: Object, writer: ITaskWriter)
...
+  const commandRegExp: RegExp = /^([^\s&"]+)(.*)$/;
+  const match: RegExpMatchArray | null = commandRegExp.exec(command);
+  if (match) {
+    // Example input: "bin/blarg --path ./config/blah.json && a/b"
+    // commandPart="bin/blarg"
+    // remainder=" --path ./config/blah.json && a/b"
+    const commandPart: string = match[1];
+    const remainder: string = match[2];
+
+    // If the command part already contains a backslash, then leave it alone
```
[the remark this answers is not in the document]
Comment (pgonzal) [reply apps/rush-lib/src/cli/taskRunner/ProjectTask.ts]: If it's something weird like that, then we err on the side of not touching their string at all.

---
In reply to: [179327633](https://github.com/Microsoft/web-build-tools/pull/601#discussion_r179327633) [](ancestors = 179327633)

File: apps/rush-lib/src/cli/taskRunner/ProjectTask.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (pgonzal) [reply apps/rush-lib/src/cli/taskRunner/ProjectTask.ts]: BTW there is a unit test showing this.

---
In reply to: [179601293](https://github.com/Microsoft/web-build-tools/pull/601#discussion_r179601293) [](ancestors = 179601293,179327756)

**Pull request merged**